### ISSUE TITLE
[Feature Request] Shared information during input transformation of models in ModelListGP

### ISSUE THREAD

**Issue by pabloprf** [pytorch/botorch#1976]
I'm trying to leverage the capabilities of botorch to implement input and outcome transformations in the models themselves, rather than externally transforming/untransforming, which is what I was doing so far.

I have the following setup:
- ModeListGP with model1 and model2.
- Both model1 and model2 have custom input transformations, which consist of step 1 and step 2.
- The two models share step 1, but have different step 2.
- Step 1 is very expensive

When I was doing the input transformation externally, I was doing step 1 only once, to avoid the extra cost of repeating the exact same thing twice. However, when using the input_transform in botorch, step 1 would be called twice, generating unnecessary computational cost at each ModelListGP evaluation.

Trying to only call step 1 once is critical for my work, because I have ~100 models inside ModelListGP and step 1 is not a negligible fraction of the cost, even when run only once.

Any ideas on how to do this? Is this a road blocker and I should go back to doing the transformation externally?

@Balandat @saitcakmak 

Thanks!

**Comment by pabloprf** [comment]
Actually, I found a reasonably simple solution for this. I just have to make sure that my input transformation function writes the results of step 1 to be accessed in the `InputTransform` class (custom one), or not to re-calculate if the variable exists already. Then I just need to modify the `transform_inputs` and `posterior` methods of `ModelListGP` to make sure that after each evaluation I delete that variable and starts fresh.
It's not the most elegant solution but it works as far as I can tell.
I don't think it's worth generalizing this, as my case might be too niche.

**Comment by saitcakmak** [comment]
Hi @pabloprf. As you found out, the solution (that doesn't require larger changes to BoTorch) is to implement some form of caching in the input transform. Instead of modifying `ModelListGP`, you could implement this in the input transform directly. A simple implementation that only caches the last evaluation (which seems to be what you need) would look something like this:

```
class MyInputTransform(InputTransform):
     last_seen: Optional[Tensor] = None
     last_transformed_value: Tensor

    def transform(self, X: Tensor) -> Tensor:
        if self.last_seen is not None and torch.equal(X, self.last_seen):
            return self.last_transformed_value
        # apply the transform as usual to get the transformed_X
        self.last_seen = X
        self.last_transformed_value = transformed_X
        return transformed_X
```

**Comment by pabloprf** [comment]
Hi @saitcakmak , thanks for the suggestion!
I will take a deeper look into this possibility, however, at first sight, it may not work for my case. The input transform is not shared among all the models in my ModelListGP, it's just that the first step in all the input transforms is the same, but the rest is different. I could write all the input transforms into a single one, by maybe passing a parameter to indicate what model it is so that it does the right thing...

**Comment by pabloprf** [comment]
Closing this issue because I found a solution that works. Leaving what I did here in case someone finds it useful, but I don't think it's worth pushing something like this to the botorch release, as it is too specific:

1. I created my own input transformation class with parent `botorch.models.transforms.input.ReversibleInputTransform` that receives a dictionary `self.surrogate_parameters`  where I will store the intermediate step 1 results. This dictionary is shared among all the models in my Model List. I also create a `self.flag_to_store` that defaults to `False`. In the `_transform` method, I make sure that I check whether results exist in the dictionary, and do not execute step 1 if they do. If `self.flag_to_store=True`, I add them to the dictionary if I have executed the expensive step 1.

2. I created a child class from `botorch.models.model_list_gp_regression.ModelListGP` and modified both the `transform_inputs` and `posterior` methods so that:
- before calling their respective `super().transform_inputs()` and `super().posterior()` I change the flag of my first model `self.models[0].input_transform.tf1.flag_to_store = True` and also delete the contents of the `self.surrogate_parameters` dictionary. 
- after calling those functions, I go back to `self.models[0].input_transform.tf1.flag_to_store = False`

This way,  I have the dictionary `self.surrogate_parameters` shared among all my models but only constructed from the first (big speed up). I also use the contents of this dictionary to accelerate the execution of the `outcome_transform`, which also uses that information as step 1.
